File: src/main/java/com/baulsupp/oksocial/ws/WebSocketPrinter.java
package com.baulsupp.oksocial.ws;

import com.baulsupp.oksocial.output.OutputHandler;
import java.util.concurrent.CountDownLatch;
import okhttp3.Response;
import okhttp3.WebSocket;
import okhttp3.WebSocketListener;
import okio.ByteString;

public class WebSocketPrinter extends WebSocketListener {
  private CountDownLatch latch = new CountDownLatch(1);
  private OutputHandler outputHandler;

  public WebSocketPrinter(OutputHandler outputHandler) {
    this.outputHandler = outputHandler;
  }

  public void waitForExit() throws InterruptedException {
    latch.await();
  }

  @Override public void onMessage(WebSocket webSocket, String text) {
    outputHandler.info(text);
  }

  @Override public void onMessage(WebSocket webSocket, ByteString bytes) {
    outputHandler.info(bytes.hex());
  }

  @Override public void onClosed(WebSocket webSocket, int code, String reason) {
    latch.countDown();
  }

  @Override public void onFailure(WebSocket webSocket, Throwable t, Response response) {
    outputHandler.showError(null, t);
    latch.countDown();
  }
}

File: src/main/java/com/baulsupp/oksocial/ws/WebSocketPrinter.kt
package com.baulsupp.oksocial.ws

import com.baulsupp.oksocial.output.OutputHandler
import java.util.concurrent.CountDownLatch
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString

class WebSocketPrinter(private val outputHandler: OutputHandler<*>) : WebSocketListener() {
    private val latch = CountDownLatch(1)

    @Throws(InterruptedException::class)
    fun waitForExit() {
        latch.await()
    }

    override fun onMessage(webSocket: WebSocket?, text: String?) {
        outputHandler.info(text)
    }

    override fun onMessage(webSocket: WebSocket?, bytes: ByteString?) {
        outputHandler.info(bytes!!.hex())
    }

    override fun onClosed(webSocket: WebSocket?, code: Int, reason: String?) {
        latch.countDown()
    }

    override fun onFailure(webSocket: WebSocket?, t: Throwable?, response: Response?) {
        outputHandler.showError(null, t)
        latch.countDown()
    }
}
